### DISCUSSION TITLE
Can SpanRuler do everything EntityRuler does?

### DISCUSSION THREAD

**Discussion by kinghuang** [explosion/spaCy#11098] [Help: Coding & Implementations] [answered]
I'm belatedly realizing that there's an `annotate_ents` config on [SpanRuler](https://spacy.io/api/spanruler) that controls whether to save spans to `doc.ents`. With that enabled, is the `SpanRuler` as functionally capable as `EntityRuler`? Is there any reason to use an `EntityRuler` over a `SpanRuler` with `annotate_ents=True`?

I'm basically wondering if I can simplify my code and use `SpanRuler` exclusively for both creating span groups and adding to `doc.ents`.

**Comment by adrianeboyd** [comment, marked as answer]
`SpanRuler` can do everything that `EntityRuler` does. If you look at the factory for `future_entity_ruler` you can see how we have implemented the original `EntityRuler` functionality using `SpanRuler`:

https://github.com/explosion/spaCy/blob/86ee26e3c29aae7a5bb56517de1e1c4ec98f41be/spacy/pipeline/span_ruler.py#L22-L63

The plan is to remove the current `EntityRuler` implementation in spacy v4 and just use `SpanRuler` underneath. `future_entity_ruler` will become `entity_ruler` so you can get the same behavior without changing your code, but we only have to maintain one component.

**Reply by kinghuang** [reply]
Awesome. The `future_entity_ruler` makes migration easy. Thanks!
